src/XMakeBuildEngine/Definition/ToolsetReader.cs

```diff
@@ -199,29 +199,32 @@ ToolsetDefinitionLocations locations
                     }
 
                     // Other toolsets are installed in the xbuild directory
-                    var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
-                    var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
-                    foreach (var d in Directory.GetDirectories(xbuildToolsetsDir).Where(d => r.IsMatch(d)))
+                    var xbuildToolsetsDir = Path.Combine(libraryPath, $"xbuild{Path.DirectorySeparatorChar}");
```

Comment:
Commit message should explain the reasoning for the changes. What is being changed can be seen in the commit itself, but the reason is important and can be useful in debugging or understanding it in future. So, something like this:

```
ToolsetReader: Escape paths before using them as part of a regex

We are trying to find toolset paths for mono, and using regex to match such
paths to a pattern. But this fails when running on windows and the path contains
`\` like `D:\j\workspace\_Windows_NT`, because `Regex` interprets it as a escape
sequence. It manifests as:

MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
11:59:04 System.ArgumentException: parsing 'D:\j\workspace\_Windows_NT_p---ab87db35\bin\xbuild\\d+\.\d+' - Unrecognized escape sequence \\j.
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanCharEscape()
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanBasicBackslash()
11:59:04    at System.Text.RegularExpressions.RegexParser.ScanRegex()
11:59:04    at System.Text.RegularExpressions.RegexParser.Parse(String re, RegexOptions op)
11:59:04    at System.Text.RegularExpressions.Regex..ctor(String pattern, RegexOptions options, TimeSpan matchTimeout, Boolean useCache)
11:59:04    at System.Text.RegularExpressions.Regex..ctor(String pattern)
11:59:04    at Microsoft.Build.Evaluation.ToolsetReader.ReadAllToolsets(Dictionary`2 toolsets, PropertyDictionary`1 environmentProperties, PropertyDictionary`1 globalProperties, ToolsetDefinitionLocations locations)
11:59:04    at Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection()
11:59:04    at Microsoft.Build.Evaluation.ProjectCollection..ctor(IDictionary`2 globalProperties, IEnumerable`1 loggers, IEnumerable`1 remoteLoggers, ToolsetDefinitionLocations toolsetDefinitionLocations, Int32 maxNodeCount, Boolean onlyLogCriticalEvents)
11:59:04    at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, Boolean debugger, Boolean detailedSummary)
11:59:04    at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)

Use `Regex.Escape` for the path portion of the pattern.

Also, see https://github.com/Microsoft/msbuild/issues/695
```